extension/source/QuillPage/QuillContext.tsx

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import getWindowQuillProvider from './getWindowQuillProvider';
+import assert from '../helpers/assert';
+import mapValues from '../helpers/mapValues';
+import { QuillInPageProvider } from '../PageContentScript/InPageProvider';
+import ExplicitAny from '../types/ExplicitAny';
+import Rpc, { rpcMap } from '../types/Rpc';
+
+type QuillContextValue = {
+  provider: QuillInPageProvider;
+  rpc: Rpc;
+};
+
+function getQuillContextValue(
+  provider: QuillInPageProvider,
+): QuillContextValue {
+  return {
+    provider,
+    rpc: {
+      public: mapValues(
+        rpcMap.public,
+        ({ params: paramsType, output }, method) => {
+          return async (...params: unknown[]) => {
+            assert(paramsType.is(params));
+            const response = await provider.request({
+              method,
+              params,
+            });
+            assert(output.is(response));
+            return response as ExplicitAny;
+          };
+        },
+      ),
+      private: mapValues(
+        rpcMap.private,
+        ({ params: paramsType, output }, method) => {
+          return async (...params: unknown[]) => {
+            assert(paramsType.is(params));
+            const response = await provider.request({
+              method,
+              params,
+            });
+            assert(output.is(response));
+            return response as ExplicitAny;
+          };
+        },
+      ),
+    },
+  };
+}
+
+const QuillContext = React.createContext<QuillContextValue>(
+  // QuillProvider render will ensure this is set properly
+  // before other components load.
+  {} as QuillContextValue,
+);
+
+export function useQuill() {
+  return React.useContext(QuillContext);
+}
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export function QuillProvider({ children }: Props) {
+  const [ctxVal, setCtxVal] = React.useState<QuillContextValue | undefined>();
+
+  React.useEffect(() => {
+    (async () => {
+      const provider = await getWindowQuillProvider();
+      const val = getQuillContextValue(provider);
+      setCtxVal(val);
+    })();
+  }, []);
+
+  if (!ctxVal) {
+    // This could be replaced by a nicer splash screen
+    return <div>Loading...</div>;
+  }
+
+  return (
+    <QuillContext.Provider value={ctxVal}>{children}</QuillContext.Provider>
+  );
+}
```

Comment:
@voltrevo This file and how it is consumed is the major refactor I did in this PR. I really didn't like delaying the ReactDOM init under an async function, so I consolidated that and other context initialization logic into this provider, turning it into more of a black box abstracting the complexity of that flow away.

Reply:
Thanks, this is useful.
Can definitely change "Loading..." with some static assets in the future.